### DISCUSSION TITLE
Adding libRAW / RAW support to libvips

### DISCUSSION THREAD

**Discussion by tru** [libvips/libvips#2075] [Ideas]
Hello,

We are evaluating vips for our project and we need to have support for RAW images (at least to read - we never output or do operations on the source file).

What would be the right way to integrate libRAW into libvips? Write a new foreign load function? Could it be modelled around something already in there like jpeg or png support?

Would such integration be welcomed upstream or would it be better if we build a pipeline in code that would convert raw to tiff before pushing it into libvips?

Thanks for reading and responding in advance.

**Comment by jcupitt** [comment]
Sure, happy to add a libraw reader. I would model it on `csvload.c`. Copy-paste that and strip out the CSV parsing and it's not too much code.

libvips can already read RAW via imagemagick. Try:

```
vips magickload panasonic.raw x.png
```

Though all IM does is shell out to `ufraw-batch` (I think).

libvips has it's own shell out system, so you can also use something like (untested):

```python
image = pyvips.Image.system(f"ufraw-batch --output=%s --out-type=ppm --out-depth=16 -i {filename}", out_format="%s.ppm")
```

**Comment by vikasjio** [comment]
@jcupitt i too would like to resize, crop etc and convert raw images to jpeg/webp.
While libvips is doing that for DNG images, i wanna do it for canon, nikon, sony, kodak etc camera raw formats like .cr2, .rw2 etc. (I tried vibvips vips-8.10.2 but it did not work for these formats)
When can we expect the same in libvips?

(i understand i might do it via dcraw converting raw to TIFF and then using libvips on TIFF but that way it take too long and i need output ideally in milliseconds and max in low single digit seconds; so that dcraw kinda multi-step approach seems unfeasible for me)

**Comment by jcupitt** [comment]
libvips is a volunteer project, so someone needs to volunteer to do the work. I don't think there's any timescale.

**Comment by jcupitt** [comment]
libraw is just a wrapper around dcraw, so I doubt if a libraw reader will be much quicker than running dcraw yourself. It might be a little faster I suppose.

**Comment by cdecker08** [comment]
has anyone taken this up since it was introduced?  if not, what would be the level of effort for a new dev to implement ourselves?

**Reply by jcupitt** [reply]
I would take a copy csvload.c and start hacking. It would only be a few 100 lines of C, though if you've not GObject or libraw before there would be a bit of a learning curve.

**Reply by cdecker08** [reply]
am i correct that csvload does not support reading metadata? is there somewhere that could be copied from?

**Reply by jcupitt** [reply]
Sure, pick another loader, perhaps `nsgifload.c`?

[vips_foreign_load_nsgif_set_header](https://github.com/libvips/libvips/blob/master/libvips/foreign/nsgifload.c#L276-L352)

The docs for the libvips metadata API are here:

https://www.libvips.org/API/current/libvips-header.html

**Comment by kleisauke** [comment]
A dedicated LibRaw loader is now available on the master branch (which will be part of the upcoming 8.18 release).
